src/backend/resources.rs

```diff
@@ -0,0 +1,228 @@
+//! Resource estimation backend for the Calyx compiler.
+//! Transforms an [`ir::Context`](crate::ir::Context) into a CSV that
+//! counts the number of different primitives in a program and loosely
+//! estimates the total size of the generated hardware.
+
+use std::io::Write;
+use std::vec;
+use std::{collections::HashMap, io};
+
+use crate::backend::traits::Backend;
+use calyx::ir::Component;
+use calyx::{
+    errors::CalyxResult,
+    ir::{self, Binding, CellType, Id},
+    utils::OutputFile,
+};
+
+#[derive(Default)]
+pub struct ResourcesBackend;
+
+impl Backend for ResourcesBackend {
+    fn name(&self) -> &'static str {
+        "resources"
+    }
+
+    /// OK to run this analysis on any Calyx program
+    fn validate(_ctx: &ir::Context) -> CalyxResult<()> {
+        Ok(())
+    }
+
+    /// Don't need to take care of this for this pass
+    fn link_externs(
+        _ctx: &ir::Context,
+        _file: &mut OutputFile,
+    ) -> CalyxResult<()> {
+        Ok(())
+    }
+
+    fn emit(ctx: &ir::Context, file: &mut OutputFile) -> CalyxResult<()> {
+        let main_comp = ctx
+            .components
+            .iter()
+            .find(|comp| comp.name == ctx.entrypoint)
+            .unwrap();
+
+        let count_map: &mut HashMap<(Id, Binding), u32> = &mut HashMap::new();
+
+        gen_count_map(ctx, main_comp, count_map);
+
+        write_csv(count_map.clone(), file);
+
+        estimated_size(count_map.clone());
+
+        Ok(())
+    }
+}
+
+/// Counts the number of each primitive with a given set of parameters
+/// in the program with entrypoint `main_comp`.
+/// TODO: handle multi-component designs
+fn gen_count_map(
+    ctx: &ir::Context,
+    main_comp: &Component,
+    count_map: &mut HashMap<(Id, Binding), u32>,
+) {
+    for cell in main_comp.cells.iter() {
+        let cell_ref = cell.borrow();
+        let primitives = vec![
+            "std_reg",
+            "std_mem_d1",
+            "std_mem_d2",
+            "std_mem_d3",
+            "std_mem_d4",
+            "seq_mem_d1",
+            "seq_mem_d2",
+            "seq_mem_d3",
+            "seq_mem_d4",
+        ];
```

Comment:
Probably the compiler is already doing this but probably worth hoisting this definition out of the for loop

Reply:
Also this could probably be an array (if you want to keep the linear search) or a hashset constructed from one, since it seems the dynamic size of the vec isn't needed here

Reply:
I actually don't 100% understand why we need to filter the primitives here to this specific set—it seems like it would be OK to include *all* primitives? (In other words, is it truly necessary to have the `if primitives.contains` check below?)

Reply:
I originally did this to make sure that the only primitives being counted were those I had size estimates for. But it definitely makes sense to just count them all. Done!